internal/rootcoord/quota_center.go

```diff
@@ -420,123 +416,200 @@ func (q *QuotaCenter) calculateWriteRates() error {
-	enableMemProtection := Params.QuotaConfig.MemProtectionEnabled.GetAsBool()
-
-	dataNodeMemoryLowWaterLevel := Params.QuotaConfig.DataNodeMemoryLowWaterLevel.GetAsFloat()
-	dataNodeMemoryHighWaterLevel := Params.QuotaConfig.DataNodeMemoryHighWaterLevel.GetAsFloat()
-	queryNodeMemoryLowWaterLevel := Params.QuotaConfig.QueryNodeMemoryLowWaterLevel.GetAsFloat()
-	queryNodeMemoryHighWaterLevel := Params.QuotaConfig.QueryNodeMemoryHighWaterLevel.GetAsFloat()
-
+	t1, _ := tsoutil.ParseTS(ts)
 	for nodeID, metric := range q.queryNodeMetrics {
-		if enableTtProtection && maxDelay >= 0 && metric.Fgm.NumFlowGraph > 0 && metric.Fgm.MinFlowGraphChannel != "" {
-			minTs, _ := tsoutil.ParseTS(metric.Fgm.MinFlowGraphTt)
-			delay := curTs.Sub(minTs)
-			processTtDelay(typeutil.QueryNodeRole, nodeID, metric.Fgm.MinFlowGraphChannel, minTs, delay, metric.Effect.CollectionIDs)
+		if metric.Fgm.NumFlowGraph > 0 && metric.Fgm.MinFlowGraphChannel != "" {
+			t2, _ := tsoutil.ParseTS(metric.Fgm.MinFlowGraphTt)
+			delay := t1.Sub(t2)
+			updateCollectionDelay(delay, metric.Effect.CollectionIDs)
+			role = fmt.Sprintf("%s-%d", typeutil.QueryNodeRole, nodeID)
+			vchannel = metric.Fgm.MinFlowGraphChannel
+			minTt = t2
 			metrics.RootCoordTtDelay.WithLabelValues(typeutil.QueryNodeRole, strconv.FormatInt(nodeID, 10)).Set(float64(delay.Milliseconds()))
 		}
-
-		if enableMemProtection {
-			processMemUsage(typeutil.QueryNodeRole, nodeID, metric.Hms, queryNodeMemoryLowWaterLevel, queryNodeMemoryHighWaterLevel, metric.Effect.CollectionIDs)
-		}
 	}
-
 	for nodeID, metric := range q.dataNodeMetrics {
-		if enableTtProtection && maxDelay >= 0 && metric.Fgm.NumFlowGraph > 0 && metric.Fgm.MinFlowGraphChannel != "" {
-			minTs, _ := tsoutil.ParseTS(metric.Fgm.MinFlowGraphTt)
-			delay := curTs.Sub(minTs)
-			processTtDelay(typeutil.DataNodeRole, nodeID, metric.Fgm.MinFlowGraphChannel, minTs, delay, metric.Effect.CollectionIDs)
+		if metric.Fgm.NumFlowGraph > 0 && metric.Fgm.MinFlowGraphChannel != "" {
+			t2, _ := tsoutil.ParseTS(metric.Fgm.MinFlowGraphTt)
+			delay := t1.Sub(t2)
+			updateCollectionDelay(delay, metric.Effect.CollectionIDs)
+			role = fmt.Sprintf("%s-%d", typeutil.DataNodeRole, nodeID)
+			vchannel = metric.Fgm.MinFlowGraphChannel
+			minTt = t2
 			metrics.RootCoordTtDelay.WithLabelValues(typeutil.DataNodeRole, strconv.FormatInt(nodeID, 10)).Set(float64(delay.Milliseconds()))
 		}
+	}
 
-		if enableMemProtection {
-			processMemUsage(typeutil.QueryNodeRole, nodeID, metric.Hms, dataNodeMemoryLowWaterLevel, dataNodeMemoryHighWaterLevel, metric.Effect.CollectionIDs)
+	collectionFactor := make(map[int64]float64)
+	for collectionID, curMaxDelay := range collectionsMaxDelay {
+		if curMaxDelay.Nanoseconds() >= maxDelay.Nanoseconds() {
+			log.RatedWarn(10, "QuotaCenter force deny writing due to long timeTick delay",
+				zap.String("node", role),
+				zap.Int64("collectionID", collectionID),
+				zap.String("vchannel", vchannel),
```

Comment:
vchannel does not necessarily belong to this collection, and printing them together may confuse.